services/src/tasks/time_estimation.rs

```diff
@@ -0,0 +1,131 @@
+use std::fmt::Display;
+
+use geoengine_operators::util::number_statistics::NumberStatistics;
+use serde::{Serialize, Serializer};
+
+#[derive(Debug, Clone, Copy)]
+pub struct TimeEstimation {
+    initial_time_stamp: std::time::Instant,
+    estimate_seconds_per_pct: NumberStatistics,
+    time_estimate_seconds: (Option<f64>, Option<f64>),
+}
+
+impl TimeEstimation {
+    pub fn new() -> Self {
+        Self {
+            initial_time_stamp: std::time::Instant::now(),
+            estimate_seconds_per_pct: NumberStatistics::default(),
+            time_estimate_seconds: (None, None),
+        }
+    }
+
+    pub fn update(&mut self, time_stamp: std::time::Instant, pct_complete: f64) {
+        let pct_complete = pct_complete.clamp(0., 1.);
+
+        let elapsed_secs = (time_stamp - self.initial_time_stamp).as_secs() as f64;
+
+        // we only update if there is progress in time and pct
+        if elapsed_secs <= 0. || pct_complete <= 0. {
+            return;
+        }
+
+        let seconds_per_pct = elapsed_secs / pct_complete;
+        self.estimate_seconds_per_pct.add(seconds_per_pct);
```

Comment:
why is this a better estimator than just taking total_elapsed_secs / total_pct_complete?

Reply:
There is no guarantee that this is better. This would mean that the last updates are more accurate than the overall time. This is also just an heuristic and could be wrong. Here, I take the overall time by the overall progress and update the estimate.